design/design-doc-agent.md

```diff
@@ -0,0 +1,665 @@
+with the same name which could cause conficts in the central cluster because all
+namespaces are rolled up into one namespace. In order to prevent conflicts, the
+agent will annotate the requirements in the central cluster with the UID of the
+namespace in the remote cluster and do the necessary checks to avoid conflicts.
+
+An illustrative installation command:
+```bash
+helm install crossplane/agent \
+  --set default-credentials-secret-name=my-sa-creds \
+  --set default-target-namespace=ns-in-central
+```
+
+While this installation time configuration provides a simple setup, it comes
+with the restriction that you cannot have the requirements in different
+namespaces with the same name. You can either try using different names or you
+can specify which namespace in the remote cluster should be synced to which
+namespace in the central cluster. In the remote cluster, `Namespace` can be
+annotated as such:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
+```
+
+The agent then will try to sync the requirements in `foo` namespace of the
+remote cluster into `bar` namespace of the central cluster instead of the
+default target namespace which is `ns-in-central` in this example. But it will
+keep using the default credentials `Secret`. In case you don't want different
+namespaces to share the same credentials `Secret`, then you can override this
+setting, too:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
+    "agent.crossplane.io/credentials-secret-name": my-other-sa-creds
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-sa-creds
+  namespace: foo
+type: Opaque
+data:
+  kubeconfig: MWYyZDFlMmU2N2Rm...
+```
+
+Now, all the requirements in the `foo` namespace of the remote cluster will be
+synced to `bar` namespace of the central cluster and all of the operations will
+be done using the credentials in `my-other-sa-creds`.
+
+#### Authorization
+
+##### Remote Cluster
+
+Since there will be one agent for the whole cluster, its own mounted
+`ServiceAccount` in that remote cluster needs to have read & write permissions
+for all of the following kinds in the remote cluster listed below:
+
+* `CustomResourceDefinition`
+* `Composition`
+* `InfrastructureDefinition`
+* `InfrastructurePublication`
+* `Secret`
+* All `*Requirement` types
+
+The last one is a bit tricky because the exact list of `*Requirement` types on
+kind level is not known during installation and it's not static; new published
+APIs should be available in the remote cluster dynamically. One option is to
+allow agent to grant `Role` and `RoleBinding`s to itself as it creates the
+necessary `CustomResourceDefinition`s in the remote cluster. However, an entity
+that is able to grant permissions to itself could greatly increase the security
+posture.
```

Comment:
Nit: "increase the security posture" reads to me like "makes it more secure", whereas I think this means "makes it less secure". Perhaps "impact the security posture" would better imply the latter?